aopy/data/bmi3d.py

```diff
@@ -564,4 +566,127 @@ def get_source_files(preproc_dir, subject, te_id, date):
+    '''
+    if df is None:
+        df = pd.DataFrame()
+
+    entries = list(zip(subjects, dates, ids))
+    for subject, date, te in tqdm(entries): 
+
+        # Load data from bmi3d hdf 
+        try:
+            data, metadata = load_preproc_exp_data(preproc_dir, subject, te, date)
+        except:
+            print(f"Entry {subject} {date} {te} could not be loaded.")
+            traceback.print_exc()
+            continue
+        event_codes = data['events']['code']
+        event_times = data['events']['timestamp']
+
+        # Trial aligned event codes and event times
+        tr_seg, tr_t = get_trial_segments_and_times(event_codes, event_times, trial_start_codes, trial_end_codes)
+
+        # Get data segments 
+        cursor_traj = get_kinematic_segments(preproc_dir, subject, te, date, trial_start_codes, trial_end_codes, 
+                                                         datatype='cursor')[0].tolist()
+        hand_traj = [None] * len(tr_seg)
+        if include_handdata:
+            hand_traj = get_kinematic_segments(preproc_dir, subject, te, date, trial_start_codes, trial_end_codes, 
+                                                  datatype='hand')[0].tolist()
+        eye_traj = [None] * len(tr_seg)
+        if include_eyedata:
+            eye_traj = get_kinematic_segments(preproc_dir, subject, te, date, trial_start_codes, trial_end_codes, 
+                                                 datatype='eye')[0].tolist()
+
+        target_idx = [code[np.isin(code, target_codes)][0] - target_codes[0] if np.sum(np.isin(code, target_codes)) == 1 else 0 for code in tr_seg]
+        target_location = get_target_locations(preproc_dir, subject, te, date, target_idx).tolist()
+
+        reward = [np.any(np.isin(reward_codes, ec)) for ec in tr_seg]
+        penalty = [np.any(np.isin(penalty_codes, ec)) for ec in tr_seg]
+        
+        df = pd.concat([df,pd.DataFrame({'subject': subject,
+                                         'te_id': te, 
+                                         'date': date, 
+                                         'event_codes': tr_seg,
+                                         'event_times': tr_t, 
+                                         'reward': reward,
+                                         'penalty': penalty,
+                                         'target_idx': target_idx,
+                                         'target_location': target_location,
+                                         'cursor': cursor_traj, 
+                                         'hand': hand_traj, 
+                                         'eye': eye_traj,
+                                        })], ignore_index=True)
+    
+    return df
+
+def concat_trials_center_out(preproc_dir, subjects, ids, dates, df=None, include_center_target=True,
+                  include_handdata=False, include_eyedata=False):
+    '''
+    Wrapper around concat_trials() specifically for center-out experiments.
+
```

Comment:
Can you include more of the assumptions of the function in the description. It seems like this is mostly just restricting the event codes used.

Reply:
Will do

Reply:
After trying out different iterations, I figured keeping track of event codes and event timestamps are an easy way to avoid mean of mean problems. This is a good intermediate step and we can use our existing behavior metrics functions on top of the event code arrays. It is also easier and flexible for data filtering through logical indexing.